src/rpc/pd/leader.rs

```diff
@@ -0,0 +1,432 @@
+        } else {
+            cluster_id = Some(cid);
+        }
+        // TODO: check all fields later?
+
+        if members.is_none() {
+            members = Some(resp);
+        }
+    }
+
+    match members {
+        Some(members) => {
+            let (client, members) = try_connect_leader(&env, security_mgr, &members, timeout)?;
+            info!("All PD endpoints are consistent: {:?}", endpoints);
+            Ok((client, members))
+        }
+        _ => Err(internal_err!("PD cluster failed to respond")),
+    }
+}
+
+fn connect(
+    env: Arc<Environment>,
+    security_mgr: &SecurityManager,
+    addr: &str,
+    timeout: Duration,
+) -> Result<(pdpb_grpc::PdClient, pdpb::GetMembersResponse)> {
+    let client = security_mgr.connect(env, addr, pdpb_grpc::PdClient::new)?;
+    let option = CallOption::default().timeout(timeout);
+    match client.get_members_opt(&pdpb::GetMembersRequest::new(), option) {
+        Ok(resp) => Ok((client, resp)),
+        Err(e) => Err(Error::Grpc(e)),
+    }
+}
+
+fn try_connect(
+    env: &Arc<Environment>,
+    security_mgr: &SecurityManager,
+    addr: &str,
+    cluster_id: u64,
+    timeout: Duration,
+) -> Result<(pdpb_grpc::PdClient, pdpb::GetMembersResponse)> {
+    let (client, r) = connect(Arc::clone(&env), security_mgr, addr, timeout)?;
+    let new_cluster_id = r.get_header().get_cluster_id();
+    if new_cluster_id != cluster_id {
+        Err(internal_err!(
+            "{} no longer belongs to cluster {}, it is in {}",
+            addr,
+            cluster_id,
+            new_cluster_id
+        ))
+    } else {
+        Ok((client, r))
+    }
+}
+
+pub fn try_connect_leader(
+    env: &Arc<Environment>,
+    security_mgr: &SecurityManager,
+    previous: &pdpb::GetMembersResponse,
+    timeout: Duration,
+) -> Result<(pdpb_grpc::PdClient, pdpb::GetMembersResponse)> {
+    let previous_leader = previous.get_leader();
+    let members = previous.get_members();
+    let cluster_id = previous.get_header().get_cluster_id();
+    let mut resp = None;
+    // Try to connect to other members, then the previous leader.
+    'outer: for m in members
+        .into_iter()
+        .filter(|m| *m != previous_leader)
+        .chain(&[previous_leader.clone()])
+    {
+        for ep in m.get_client_urls() {
+            match try_connect(&env, security_mgr, ep.as_str(), cluster_id, timeout) {
+                Ok((_, r)) => {
+                    resp = Some(r);
+                    break 'outer;
+                }
+                Err(e) => {
+                    error!("failed to connect to {}, {:?}", ep, e);
+                    continue;
+                }
+            }
+        }
+    }
+
+    // Then try to connect the PD cluster leader.
+    if let Some(resp) = resp {
+        let leader = resp.get_leader().clone();
+        for ep in leader.get_client_urls() {
+            if let Ok((client, r)) =
+                try_connect(&env, security_mgr, ep.as_str(), cluster_id, timeout)
```

Comment:
Better to use an explicit return here since you do on the one below as well. :)

Reply:
I'm not sure if I understand it correctly. Can you explain it a little bit more in case I misunderstood. Thanks.

Reply:
I thought client might no be able to connect to leader's all client urls, so it tries to connect each of them and return the connection if it succeeds. So maybe it should be something like this?
```rust
    // Then try to connect the PD cluster leader.
    if let Some(resp) = resp {
        let leader = resp.get_leader().clone();
        for ep in leader.get_client_urls() {
            let r = try_connect(&env, security_mgr, ep.as_str(), cluster_id, timeout);
            if r.is_ok() {
                return r;
            }   
        }   
    } 
```

Reply:
I misread, it's fine :)